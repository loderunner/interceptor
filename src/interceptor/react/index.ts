export { Provider as InterceptProvider, useIntercepts } from './intercept'
