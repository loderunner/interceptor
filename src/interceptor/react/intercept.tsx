import * as React from 'react'
import { createContext, useContext, useState } from 'react'

import * as intercept from '../intercept'

interface InterceptContextType {
  intercepts: intercept.Intercept[]
  addIntercept: (pattern: string) => void
  removeIntercept: (inter: intercept.Intercept) => void
}

const InterceptContext = createContext<InterceptContextType>({
  intercepts: [],
  addIntercept: () => {},
  removeIntercept: () => {},
})

interface Props {
  children: React.ReactNode
}

export const Provider = ({ children }: Props) => {
  const [intercepts, setIntercepts] = useState([...intercept.intercepts])

  const addIntercept = (pattern: string) => {
    intercept.addIntercept({ pattern, enabled: true })
    setIntercepts([...intercept.intercepts])
  }
  const removeIntercept = (inter: intercept.Intercept) => {
    intercept.removeIntercept(inter)
    setIntercepts([...intercept.intercepts])
  }

  return (
    <InterceptContext.Provider
      value={{ intercepts, addIntercept, removeIntercept }}
    >
      {children}
    </InterceptContext.Provider>
  )
}

export const useIntercepts = () => useContext(InterceptContext)
