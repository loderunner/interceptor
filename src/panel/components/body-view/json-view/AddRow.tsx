import * as React from 'react'
import { useMemo, useState } from 'react'

import { Clear as ClearIcon } from '@/icons'

import KeyView from './KeyView'
import PrimitiveView from './PrimitiveView'

import type { JSONValue } from '../JSON'

interface Props {
  depth: number
  keyEditable: boolean
  onAdd?: (key: string, value: JSONValue) => void
  onCancel?: () => void
}

const AddRow = ({ depth, keyEditable, onAdd, onCancel }: Props) => {
  const [step, setStep] = useState<'key' | 'value'>(
    keyEditable ? 'key' : 'value'
  )
  const [key, setKey] = useState('')

  const keyView = useMemo(
    () => (
      <KeyView
        name={key}
        editable={keyEditable}
        editingInitial={step === 'key'}
        onChange={(k) => {
          setKey(k)
          setStep('value')
        }}
        onCancel={onCancel}
      />
    ),
    [key, keyEditable, onCancel, step]
  )

  const valueView = useMemo(
    () => (
      <PrimitiveView
        editingInitial={step === 'value'}
        onChange={(value) => {
          if (onAdd !== undefined) {
            onAdd(key, value)
          }
        }}
        onCancel={onCancel}
      />
    ),
    [key, onAdd, onCancel, step]
  )

  return (
    <div
      className="mt-1 flex items-center space-x-1"
      style={{ marginLeft: `${depth}rem` }}
    >
      {/* Placeholder icon for alignment */}
      <ClearIcon className="opacity-0 fill-white" />
      {keyView}
      {step === 'value' ? valueView : null}
    </div>
  )
}

export default AddRow
