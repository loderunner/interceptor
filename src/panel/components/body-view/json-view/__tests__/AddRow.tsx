import { cleanup, render } from '@testing-library/react'
import * as React from 'react'
import '@testing-library/jest-dom'

import AddRow from '../AddRow'

describe('[JSON.AddRow]', () => {
  afterEach(() => {
    cleanup()
  })

  it('should match initial snapshot', () => {
    const { container } = render(<AddRow depth={0} />)
    expect(container).toMatchSnapshot()
  })
})
