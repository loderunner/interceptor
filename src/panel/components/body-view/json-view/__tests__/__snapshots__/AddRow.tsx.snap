// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[JSON.AddRow] should match initial snapshot 1`] = `
<div>
  <div
    class="mt-1 flex items-center space-x-1"
    style="margin-left: 0rem;"
  >
    <svg
      class="opacity-0 fill-white"
    />
    <pre
      class="text-amber-700"
    >
      
      :
    </pre>
     
    <div
      class="flex-auto"
    >
      <pre
        class="whitespace-pre-wrap break-all "
      >
        â€‹
      </pre>
      <input
        class="fixed dark:bg-slate-600 dark:text-slate-100"
        style="padding: 0.25rem; left: 0px; top: 0px; width: calc(0px + 2 * 0.25rem); min-width: 120px;"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
